File: bindings/pyomq/examples/zguide/03_pipeline/ventilator.py
"""ZGuide 03 — Ventilator (task producer).

PUSH socket binds and sends tasks followed by END sentinels so each
worker knows when to stop.

    python ventilator.py [vent_ep] [n_tasks] [n_workers]
"""
import sys
import time

import pyomq as zmq

vent_ep = sys.argv[1] if len(sys.argv) > 1 else "ipc://@omq-zguide-03-ventilator"
n_tasks = int(sys.argv[2]) if len(sys.argv) > 2 else 100
n_workers = int(sys.argv[3]) if len(sys.argv) > 3 else 3

with zmq.Context() as ctx:
    push = ctx.socket(zmq.PUSH)
    push.bind(vent_ep)
    time.sleep(0.2)

    for i in range(n_tasks):
        push.send_string(f"task-{i}")

    for _ in range(n_workers):
        push.send(b"END")

    print(f"ventilator: sent {n_tasks} tasks + {n_workers} END sentinels on {vent_ep}")

File: examples/zguide/03_pipeline/ventilator.rs
//! `ZGuide` 03 — Ventilator (task producer).
//!
//! PUSH socket binds and sends N tasks. Workers are killed by the
//! run script after the sink has collected all results.
//!
//!     cargo run -p zguide-03-pipeline --bin ventilator [vent_ep] [n_tasks]

use std::time::Duration;

use omq_tokio::{Context, Endpoint, Message, Options, SocketType};

fn endpoint_or(args: &[String], index: usize, default: &str) -> Endpoint {
    args.get(index).map_or_else(
        || default.parse().unwrap(),
        |s| s.parse().expect("invalid endpoint"),
    )
}

#[tokio::main(flavor = "current_thread")]
async fn main() {
    let ctx = Context::new();
    let args: Vec<String> = std::env::args().collect();
    let vent_ep = endpoint_or(&args, 1, "ipc://@omq-zguide-03-ventilator");
    let n_tasks: usize = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(1000);

    let push = ctx.socket(
        SocketType::Push,
        Options::default().linger(Duration::from_secs(2)),
    );
    push.bind(vent_ep.clone()).await.unwrap();

    tokio::time::sleep(Duration::from_millis(500)).await;

    for i in 0..n_tasks {
        push.send(Message::single(format!("task-{i}")))
            .await
            .unwrap();
    }

    println!("ventilator: sent {n_tasks} tasks on {vent_ep}");
    push.close().await.unwrap();
}
